File: Day045_2025/Q045_GFG.java
class Q045_GFG {
    Node first = null, middle = null, last = null, prev = null;
    private void getWrongNodes(Node root) {
        if (root == null) 
            return;

        getWrongNodes(root.left);

        if (prev != null && root.data < prev.data) {
            if (first == null) {
                first = prev;
                middle = root;
            } else {
                last = root;
            }
        }
        
        prev = root;

        getWrongNodes(root.right);
    }

    public void correctBST(Node root) {
        
        getWrongNodes(root);

        if (first != null && last != null) {
            // Swap the data of the first and last nodes
            int temp = first.data;
            first.data = last.data;
            last.data = temp;
        } 
        else if (first != null && middle != null) {
            // Swap the data of the first and middle nodes
            int temp = first.data;
            first.data = middle.data;
            middle.data = temp;
        }
    }
}

File: Day045_2025/Q045_GFG.py
class Solution:
    def getWrongNodes(self, root, first, middle, last, prev):
        if root is None:
            return
        
        self.getWrongNodes(root.left, first, middle, last, prev)
        
        if prev and root.data < prev.data:
            if first is None:
                first = prev
                middle = root
            else:
                last = root
                
        prev = root
        self.getWrongNodes(root.right, first, middle, last, prev)
    
    def correctBST(self, root):
        first, middle, last, prev = None, None, None, None
        self.getWrongNodes(root, first, middle, last, prev)
        
        if first and last:
            # Swap the data of the first and last nodes
            first.data, last.data = last.data, first.data
        elif first and middle:
            # Swap the data of the first and middle nodes
            first.data, middle.data = middle.data, first.data
